src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs

```diff
@@ -134,49 +139,62 @@ private void AddTagsForLargeNumberOfSpans(NormalizedSnapshotSpanCollection reque
 
             using var _1 = SegmentedListPool.GetPooledList<ITagSpan<TTag>>(out var tempList);
 
-            AddIntersectingSpans(mergedSpan, tempList);
+            AppendIntersectingSpansInSortedOrder(mergedSpan, tempList);
             if (tempList.Count == 0)
                 return;
 
+            // Note: both 'requstedSpans' and 'tempList' are in sorted order.
+
             using var enumerator = tempList.GetEnumerator();
 
             if (!enumerator.MoveNext())
                 return;
 
             using var _2 = PooledHashSet<ITagSpan<TTag>>.GetInstance(out var hashSet);
 
+            var requestIndex = 0;
             while (true)
             {
-                var requestIndex = 0;
                 var currentTag = enumerator.Current;
 
                 var currentRequestSpan = requestedSpans[requestIndex];
                 var currentTagSpan = currentTag.Span;
 
-                if (currentRequestSpan.Start > currentTagSpan.End)
+                // The current tag is *before* the current span we're trying to intersect with.  Move to the next tag to
+                // see if it intersects with the current span.
+                if (currentTagSpan.End < currentRequestSpan.Start)
                 {
+                    // If there are no more tags, then we're done.
                     if (!enumerator.MoveNext())
-                        break;
+                        return;
```

Comment:
thought it was clearer to indicate the entire algorithm is done, as opposed to breaking the loop, which then exits the method.